src/ic-cdk/src/api/call.rs

```diff
@@ -241,6 +241,72 @@ fn cleanup(state_ptr: *const InnerCell<CallFutureState<Vec<u8>>>) {
     }
 }
 
+fn add_payment(payment: u128) {
+    if payment == 0 {
+        return;
+    }
+    let high = (payment >> 64) as u64;
+    let low = (payment & u64::MAX as u128) as u64;
+    unsafe {
+        ic0::call_cycles_add128(high as i64, low as i64);
+    }
+}
+
+/// Sends a one-way message that invokes `method` with arguments `args` on the principal
+/// identified by `id`, ignoring the reply.
+///
+/// Returns `Ok(())` if the message was successfully enqueued, otherwise returns a reject code.
+///
+/// # Notes
+///
+///   * The caller has no way of checking whether the destination processed the notification.
+///     The system can drop the notification if the destination does not have resources to
+///     process the message (for example, if it's out of cycles or queue slots).
+///
+///   * The callee cannot tell whether the call is one-way or not.
+///     The callee must produce replies for all incoming messages.
+///
+///   * It is safe to upgrade a canister without stopping it first if it sends out *only*
+///     one-way messages.
+pub fn notify<T: ArgumentEncoder>(
```

Comment:
Shouldn't this be called `notify_with_payment` to be consistent with `call_*`? Then there can be a `notify` method that sets the `payment` to 0 like `call` does for `call_raw`.

Reply:
Maybe it's worth it to introduce a builder API for calls.

Reply:
If you only have a few variants, builders are more code for the implementor *and* more code for the user without much benefit.

Reply:
Renamed this function to `notify_with_payment128` and added `notify` without payment for consistency.